### ISSUE TITLE
Backward() method of Convolution layer fails

### ISSUE THREAD

**Issue by hello-fri-end** [mlpack/mlpack#2986]
#### Simple code to reproduce the error
```
  arma::mat input, output, delta;
  input.ones(36,1);
  Convolution<> module(1, 1, 3, 3, 2, 2, 0, 0, 6, 6);
  module.Reset();
  module.Forward(input, output);
  module.Backward(input, output, delta);
```
#### Output

```
error: addition: incompatible matrix dimensions: 6x6 and 5x5
terminate called after throwing an instance of 'std::logic_error'
  what():  addition: incompatible matrix dimensions: 6x6 and 5x5
Aborted (core dumped)

```
@zoq, I think this is related to why the test for Inception3B fails in #2963.

**Comment by zoq** [comment]
Agreed, thanks for the simple snippet makes it easy to reproduce.

**Comment by zoq** [comment]
When I use https://github.com/mlpack/mlpack/blob/98e8e832abf5bf5e3362da06d66c23b1ba703c8d/src/mlpack/methods/ann/convolution_rules/naive_convolution.hpp instead of the current implementation it works, so it looks like there is something we are doing wrong with the dilation parameter.

**Comment by hello-fri-end** [comment]
> When I use https://github.com/mlpack/mlpack/blob/98e8e832abf5bf5e3362da06d66c23b1ba703c8d/src/mlpack/methods/ann/convolution_rules/naive_convolution.hpp instead of the current implementation it works, so it looks like there is something we are doing wrong with the dilation parameter.

Can you post the output of `delta.print()` ? I think delta will have some garbage values even in the commit you posted.

**Comment by hello-fri-end** [comment]
For the snippet I posted, the [if condition](https://github.com/mlpack/mlpack/blob/master/src/mlpack/methods/ann/convolution_rules/naive_convolution.hpp#L113) is never true. The size of `outputRows` & ` outputCols` from the [initial expressions](https://github.com/mlpack/mlpack/blob/master/src/mlpack/methods/ann/convolution_rules/naive_convolution.hpp#L106-L109) comes out to be (7,7) but it needs to be (8,8) for `ValidConvolution` to give the correct output shape as (6,6). So, I'm guessing there's something wrong with the if condition.

**Comment by rcurtin** [comment]
@rcurtin you should check to see if this test still fails on the `ann-vtable` branch!  (And, if it does, you should fix it!  (And, regardless of how the branch currently performs, you should add this as a test case!))

**Comment by hello-fri-end** [comment]
@rcurtin this snippet still fails on the `ann-vtable` branch. It still gives the same error:

```
  {Unknown expression after the reported line}
due to unexpected exception with message:
  addition: incompatible matrix dimensions: 6x6 and 5x5
```
Here's the [link](https://github.com/hello-fri-end/mlpack-zoq/blob/ann-vtable/src/mlpack/tests/ann_layer_test.cpp#L3741) for the modified snippet,

**Comment by rcurtin** [comment]
I believe I just fixed this in #2777.  It seems as through the backwards pass should use a stride of 1, but a dilation factor equal to the forwards pass stride.
